### DISCUSSION TITLE
Could I use ESPNet2 in Unity?

### DISCUSSION THREAD

**Discussion by TomohiroHibino600** [espnet/espnet#4652] [Q&A] [answered]
I would be very grateful if we could ask when you are available.

I'm currently developing in Unity and are looking for a way to enable ASR / TTS in an offline environment.
Is it possible to use ESPNet2 in Unity (C#)? If so, what would be the appropriate method?

I know that there is the following implementation for VOSK, but I could not find it for ESPNet2.
Since I am a beginner engineer, I am sorry that I cannot implement it myself.
https://assetstore.unity.com/packages/tools/audio/recognissimo-cross-platform-offline-speech-recognition-203101
https://github.com/Yeti47/Vosk4Unity

Thank you very much in advance for your time.
If this is an inappropriate issue, you may close it.

**Comment by b-flo** [comment, marked as answer]
Hi,

I guess the best way would be to export your model to ONNX and rely on [Barracuda](https://github.com/Unity-Technologies/barracuda-release) or something similar. 

> Since I am a beginner engineer, I am sorry that I cannot implement it myself.

Not sure there are an out-of-the-box solution, some rewriting in C# will be needed from my understanding.

P.S: I'm moving the issue to Discussion. It's out of scope for this repo but the discussion may be of interest for other people.

**Reply by b-flo** [reply]
Set as answer for now. If someone as a better solution, I'll change the tagged answer.

**Comment by TomohiroHibino600** [comment]
Thank you so much for your thoughtful response!

I have not succeeded in using ESPNet with Unity, but I have tried a few things and will share the information. 
Anyone do not need to reply to me.

But, I have not been able to do it properly because there are so many things I don't understand. 
Sorry if the information is not helpful. You guys are really amazing at understanding machine learning...!

1) Output the model in ONNX
I was able to output the model in ONNX format from the following
At this time, args may be necessary, but I didn't specify any for now.
https://pypi.org/project/espnet-onnx/0.1.7/

　　Tag names were specified with reference to the following
　　[espnet/Shinji_Watanabe_laborotv_asr_train_asr_conformer2_latest33_raw_char_sp_valid.acc.ave](https://huggingface.co/espnet/Shinji_Watanabe_laborotv_asr_train_asr_conformer2_latest33_raw_char_sp_valid.acc.ave)
　　[espnet/kan-bayashi_tsukuyomi_tts_finetune_full_band_jsut_vits_raw_phn_jaconv_pyopenjtalk_prosody_latest](https://huggingface.co/espnet/kan-bayashi_tsukuyomi_tts_finetune_full_band_jsut_vits_raw_phn_jaconv_pyopenjtalk_prosody_latest)

2) Import in Unity
I could not import some ONNX files into my Unity project. ONNX files with a blank icon are the ones that failed to import.
![ONNXFiles](https://user-images.githubusercontent.com/62923803/193935845-1533acef-d52c-4246-ae2f-d53e7023f7be.png)
![BlankONNX](https://user-images.githubusercontent.com/62923803/193936634-ee057063-ee55-4cb6-bf3c-d76cf9b8224f.png)

　　These files may have unsupported operations in Unity, or they may need to be edited in some way.

3) Edit in Unity
I then tried to use the files I was able to import in Barracuda. 
However, I gave up due to lack of knowledge on how to ASR / TTS using the output from the model.
https://docs.unity3d.com/Packages/com.unity.barracuda@3.0/manual/index.html

**Reply by b-flo** [reply]
Some notes:

- Make sure the task/model/architecture you're trying to use is supported by ESPnet-ONNX. Not everything is covered yet.
- Some operations may be unsupported by Barracuda, you'll have to take a look at the doc or repo.
- It seems that [onnxuntime](https://github.com/microsoft/onnxruntime) can be used in Unity so Barracuda may not be needed.

**Reply by Masao-Someki** [reply]
Hi @TomohiroHibino600, as for the ASR model, ONNX models are separated into some models, such as encoder, decoder, and ctc.
So we need to create an ASR system in csharp to use the onnx models...
Since implementing an ASR system is a hard work, I recommend you start with the TTS model.
You only need one `.onnx` file to run TTS (VITS in your screenshot), and I think the implementation will be similar to the tutorial of the [onnxruntime-c#](https://onnxruntime.ai/docs/get-started/with-csharp.html)
The required argument or supported export configuration for VITS is written [here](https://github.com/espnet/espnet_onnx/blob/master/docs/TTSModelDetail.md)

**Reply by TomohiroHibino600** [reply]
I got as far as loading vits model in my Unity project using [onnxruntime-c#](https://onnxruntime.ai/docs/get-started/with-csharp.html)!

I am getting the following error, and I believe I am not correctly inputting to the model, but this is a temporary record.
I will look into the documentation, etc.

I will share the information when I could run TTS.
Thank you all!

```
OnnxRuntimeException: [ErrorCode:RuntimeException] Non-zero status code returned while running ReduceMax node. 
Name:'ReduceMax_137' 
Status Message: D:\a\_work\1\s\engine\lotus\include\onnxruntime\core/framework/op_kernel_context.h:40 
onnxruntime::OpKernelContext::Input Missing Input: text_length

Microsoft.ML.OnnxRuntime.InferenceSession.Run
```

**Comment by sw005320** [comment]
It sounds a very cool attempt.
I just added @Masao-Someki in this discussion, who is developing espnet_onnx

**Reply by sw005320** [reply]
@Masao-Someki, it would be great if you also join the discussion.
